server/service/service.go

```diff
@@ -58,13 +58,54 @@ type SecurityArtifacts struct {
 	TLSKeypair *tls.Certificate
 }
 
-// EntityLookup provides a way to resolve chassis and control cards
-// in the EntityManager.
+// EntityLookup is used to resolve the fields of an active control card to a chassis.
+// For fixed form factor devices, the active control card is the chassis itself.
 type EntityLookup struct {
+	// The manufacturer of this control card or chassis.
 	Manufacturer string
+	// The serial number of this control card or chassis.
 	SerialNumber string
-	PartNumber   string
-	IPAddress    string
+	// The hardware model/part number of this control card or chassis.
+	PartNumber string
+	// The reported IP address of the management interface for this control
+	// card or chassis.
+	IPAddress string
+	// Whether this chassis appears to be a modular device.
+	Modular bool
+}
+
+// buildEntityLookup constructs an EntityLookup object from a bootstrap request.
+func buildEntityLookup(ctx context.Context, req *bpb.GetBootstrapDataRequest) (*EntityLookup, error) {
+	peerAddr, err := peerAddressFromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+	activeControlCardSerial := req.GetControlCardState().GetSerialNumber()
+	var partNumber string
+	var modular bool
+	if len(req.GetChassisDescriptor().GetControlCards()) == 0 {
+		modular = false
+		partNumber = req.GetChassisDescriptor().GetPartNumber()
+	} else {
+		modular = true
+		for _, card := range req.GetChassisDescriptor().GetControlCards() {
+			if card.GetSerialNumber() == activeControlCardSerial {
+				partNumber = card.GetPartNumber()
+				break
+			}
+		}
+	}
+	if partNumber == "" {
+		return nil, status.Errorf(codes.InvalidArgument, "active control card with serial %v not found in chassis descriptor", activeControlCardSerial)
```

Comment:
NotFound?

Reply:
NotFound implies that the server could not find something (e.g. server's fault). I think invalid argument is more correct here as it is actually the user/client that hasn't filled in the request properly.

Reply:
oke sgtm